python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1375,20 +1378,39 @@ def _validate_self_deps(
     input_keys: Iterable[AssetKey],
     output_keys: Iterable[AssetKey],
     partition_mappings: Mapping[AssetKey, PartitionMapping],
+    partitions_def: Optional[PartitionsDefinition],
 ) -> None:
+    is_valid_self_dep_mapping = (
+        lambda partition_mapping: isinstance(partition_mapping, TimeWindowPartitionMapping)
+        and (partition_mapping.start_offset or 0) < 0
+        and (partition_mapping.end_offset or 0) < 0
+    )
     output_keys_set = set(output_keys)
     for input_key in input_keys:
         if input_key in output_keys_set:
             if input_key in partition_mappings:
                 partition_mapping = partition_mappings[input_key]
                 if (
-                    isinstance(partition_mapping, TimeWindowPartitionMapping)
-                    and (partition_mapping.start_offset or 0) < 0
-                    and (partition_mapping.end_offset or 0) < 0
+                    isinstance(partitions_def, TimeWindowPartitionsDefinition)
```

Comment:
```
if time_window_partitions_def is not None:
   time_window_partition_mapping = get_time_window_partition_mapping(partition_mapping)
   if (
       time_window_partition_mapping is not None
       and (partition_mapping.start_offset or 0) < 0
       and (partition_mapping.end_offset or 0) < 0
   ):
       continue
 ```
 
 where those proposed functions handle the mess of sorting out the different cases (partitions_def is None, partitions_def is TimeWindow, partitions_def is Multi with no time dimension, etc.)
 
 this way we don't have to duplicate this work in a bunch of callsites